clr/commands.py

```diff
@@ -74,44 +76,158 @@ def resolve_command(query):
+    signature = Signature.from_callable(command_callable)
+
+    if signature.return_annotation != Signature.empty:
+        print(f'WARNING: {command_callable} returns a {signature.return_annotation} which is ignored.')
+    for param in signature.parameters.values():
+        if param.kind == param.VAR_KEYWORD:
+            print(f'WARNING: Ignoring kwargs found for clr command {param} {command_callable}: {varkwarg}')
 
     if args is None:
         args = tuple()
     if defvals is None:
         defvals = tuple()
 
     # Avoid the self argument.
-    if isinstance(cmd, types.MethodType):
+    if isinstance(command_callable, types.MethodType):
         args = args[1:]
 
     nargs = len(args) - len(defvals)
     args = list(zip(args[:nargs], [NO_DEFAULT]*nargs)) + list(zip(args[nargs:], defvals))
+    return CommandSpec(args, vararg, inspect.getdoc(command_callable))
+
+
+@dataclass
+class Namespace:
+    descr: str
+    longdescr: str
+    command_specs: dict
+    command_callables: dict
+
+    @property
+    def commands(self):
+        return sorted(self.command_specs.keys())
+
+@dataclass
+class ErrorLoadingNamespace:
+    """Psuedo namespace for when one can't be loaded to show the error message."""
+    key: str
+    error: Exception
+
+    commands = {}
+    command_specs = {}
+
+    @property
+    def descr(self):
+        return f"ERROR Could not load. See `clr help {self.key}`"
+
+    @property
+    def longdescr(self):
+        return f"Error importing module '{clr.config.commands()[self.key]}' for namespace '{self.key}':\n\n{self.error}"
+
+@dataclass(frozen=True)
+class NamespaceCacheEntry:
+    descr: str
+    longdescr: str
+    command_specs: dict
 
-    return args, vararg
+    @staticmethod
+    def create(namespace):
+        return NamespaceCacheEntry(namespace.descr, namespace.longdescr, namespace.command_specs)
+
+    @property
+    def commands(self):
+        return sorted(self.command_specs.keys())
+
+class NamespaceCache:
+    """Cache introspection on command names and signatures to disk.
+
+    This allows subsequent calls to `clr help` or `clr completion` to be fast.
+    Necessary to work around the fact that many clr command namespace modules
+    import the world and initialize state on import.
+    """
+
+    def __init__(self):
+        tmpdir = os.environ.get('TMPDIR', '/tmp')
+        self.CACHE_FN = os.path.join(tmpdir, 'clr_command_cache')
+        # Clr processes are short lived. We don't close the shelve, but are
+        # careful to sync it after writes.
+        self.cache = shelve.open(self.CACHE_FN)
+
+    def get(self, namespace_key):
+        # Don't cache the system namespace. It is already loaded.
+        if namespace_key == 'system': return get_namespace('system')
+        if namespace_key not in self.cache:
+            namespace = get_namespace(namespace_key)
+            if isinstance(namespace, ErrorLoadingNamespace): return namespace
```

Comment:
I realize that one _can_ do one-line returns like this, but most Color source prefers the explicit linebreak after control flow.

I guess the `clr` repo isn't set up with either `flake8` or `pylint` or anything, but I think those will flag this stuff.